sp/src/game/shared/hl2/weapon_css_base.cpp

```diff
@@ -128,7 +128,17 @@ void CBase_CSS_HL2_Pistol::FireNPCPrimaryAttack( CBaseCombatCharacter *pOperator
 	CSoundEnt::InsertSound( SOUND_COMBAT|SOUND_CONTEXT_GUNFIRE, pOperator->GetAbsOrigin(), SOUNDENT_VOLUME_PISTOL_CSS, 0.2, pOperator, SOUNDENT_CHANNEL_WEAPON, pOperator->GetEnemy() );
 
 	WeaponSound( SINGLE_NPC );
-	pOperator->FireBullets( 1, vecShootOrigin, vecShootDir, VECTOR_CONE_PRECALCULATED, MAX_TRACE_LENGTH, m_iPrimaryAmmoType, 2, -1, -1, round( GetAmmoDef()->NPCDamage( m_iPrimaryAmmoType ) * GetNPCDamageMultiplier() ) );
+
+	FireBulletsInfo_t info( 1, vecShootOrigin, vecShootDir, VECTOR_CONE_PRECALCULATED, MAX_TRACE_LENGTH, m_iPrimaryAmmoType );
+	info.m_iTracerFreq = 2;
+
+	if (GetNPCDamageMultiplier() != 1.0f)
+	{
+		info.m_flDamage = round( GetAmmoDef()->NPCDamage( info.m_iAmmoType ) * GetNPCDamageMultiplier() );
+		info.m_nFlags |= FIRE_BULLETS_NO_AUTO_GIB_TYPE;
```

Comment:
I feel a little confused here. What is the check for ``GetNPCDamageMultiplier() != 1.0f`` for? Why does it only specify no auto gib if the NPC damage multiplier isn't 1.0?

Reply:
`FIRE_BULLETS_NO_AUTO_GIB_TYPE` only comes into play when `m_flDamage` is set to a non-zero value. It's meant to bypass a hardcoded feature in which any `m_flDamage` value greater than `20` adds `DMG_ALWAYSGIB` to the bullet's damage types.

Reply:
I still don't understand. If the result of ``GetNPCDamageMultiplier()`` is 1.0f, that means that m_flDamage should also be 1.0? But what if ``GetAmmoDef()->NPCDamage( info.m_iAmmoType)`` is not 1? This does not look right to me.

Reply:
No, `m_flDamage` should not be 1.0. It is a direct damage override value, not a multiplier.

When `m_flDamage` is set to 0 (the default), the bullet firing code uses the ammo type's player or NPC damage, depending on which type of entity is firing the bullets. When `m_flDamage` is not set to 0, the bullet firing code overrides those values and uses the value stored in `m_flDamage` instead.

`GetNPCDamageMultiplier()` is intended to be multiplied with the ammo type's NPC damage, so it takes the ammo type's NPC damage in the same manner as the underlying code, multiplies it, rounds it to the nearest whole number, and then assigns that value to `m_flDamage` to override the default behavior. When `GetNPCDamageMultiplier()` is set to 1.0, the damage does not need to be overridden because the default NPC damage value multiplied by 1.0 is equivalent to the default NPC damage value.

Reply:
I'm sorry, I didn't understand the default behavior. This makes perfect sense now.

One last question - why is DMG_ALWAYSGIB only applied if m_flDamage is not passed as 0? It seems weird to me that the default damage doesn't cause gibbing but override damage does

Reply:
I don't know. `m_flDamage` only seems to be used by `func_tank` in default HL2, so I suspect this is from some kind of unused legacy case. The flag `FIRE_BULLETS_NO_AUTO_GIB_TYPE` is one I added in a previous Mapbase update, and it's meant to get around this functionality without interfering with default behavior.